### ISSUE TITLE
music_transformer issus

### ISSUE THREAD

**Issue by firedie** [bearpelican/musicautobot#4]
Thank you for your repo, it is brilliant!
I am new to this area and I have got several questions
1 I notitced the CLI scripts both for multitask training and music transformer training are runing run_multitask.py, so did you just discard run_music_transformer.py? If not, when I try to run this file, it pops several errors, are you gonna to fix that?
2 Is about the notebook scripts, I noticed that during the generate procedure, you still need to load the pretrained data pkl file , is that mean we need the pretrain data as well as the pretrained model to help us make prediction?
3. When I run the multitask_transformer/Generate.ipynb and to the line of:
pred_nw = learn.predict_nw(seed, n_words=200)
sometimes it shows it has been interrupted but if I try several times, finally it will work, any reason for that?
4 I try to use my own dataset to train, is there any restriction for the format, like minimum bas numbers? Does it has to be 2 tracks?
5 I try to train with your example midi files, sometime the prediction result will be empty, any suggestion for that?

Truly thankful for your reply!

**Comment by firedie** [comment]
more over, when I try to do some training some error always pop out:
..\aten\src\ATen\native\LegacyDefinitions.cpp:14: UserWarning: masked_fill_ received a mask with dtype torch.uint8, this behavior is now deprecated,please use a mask with dtype torch.bool instead.
Do you have the same warning or thats my setting issue

**Comment by bearpelican** [comment]
0bee0e6e7cc4f23a66a31a798ef882de4098ef7b should fix `masked_fill_ received a mask with dtype torch.uint8`.  This was a pytorch version mismatch

**Comment by bearpelican** [comment]
1. Oops thanks for catching the bug! I just updated the README. The second command should be `run_music_transformer.py` instead of `run_multitask.py`

2. You can use the pretrained model or one that you have pretrained yourself to make a prediction. The `data` is whatever you preprocessed in the `Train.ipynb` notebook. So you'll need to run that first

3. Can you print out the logs when it gets interrupted? The prediction function will stop short if it encounters an EOS token. Meaning it has predicted the end of the song. That might be why you see it stopping short of the `n_words` parameter.

4. The `music_transformer` model has no restrictions on dataset. The `multitask_model` currently requires midi files with only 2 tracks. I'll be working on removing that restriction in the future.

5. Yeah I'm afraid you'll need to collect more MIDI files in order to get better predictions. I only have a few files available because I don't completely own the data

**Comment by firedie** [comment]
Thanks for you reply!

About question No.2, is that means when I made a prediction, I have to get both of my pre-train data i.e. pkl file as well as pretrain model i.e. pth file ready and the pre-train model should match the pre-train data so that I can get some reasonable result?
As for my own knowledge, the pre-train model should be enough so I am quite confuesed about what is the data here.
Moreover, I try to download your pre-train model and combining with my own data set, the result is unsatisfied, espicially for the next word generation, it is basically repeating the first two bar with minor changes. (Although I know you have implement top-p top-k sampling to prevent such a problem however bad thing happened :( )

As for the training process, how long or how many epochs does it take you to do the training to get some good result? And could share the numbers of how many midi files did you get? I am trying to get some midi files myself and not quite sure how much is 'enough'

Cheers

**Comment by bearpelican** [comment]
Ah yes. For No. 2, you don't really need the pre-trained data. We don't use any of it for generation. 
However, it's one of the required parameters to create a `Learner` in fastai. You could probably create an empty dataset and pass it in too.

Repeating notes:
Have you tried increasing temperature?
The default settings I use for musicautobot are:
Temperature (Pitch, Duration) = (1.2, 0.8) 
TopP = 0.9
TopK = 24

I had quite a bit of MIDI files (180k). 
The current models are specifically trained in the key of C. I'll be adding a few more models over the weekend. (Smaller and key agnostic)
I'd suggest starting from one of those and fine tuning on your data. Or if you need to train from scratch, I'd use the the [Reddit](https://www.reddit.com/r/datasets/comments/3akhxy/the_largest_midi_collection_on_the_internet/) dataset for a generalized model. Then fine tune on your target MIDI from there.

**Comment by firedie** [comment]
![image](https://user-images.githubusercontent.com/18274020/64140473-19a0c300-ce37-11e9-8016-e117009a0121.png)
I try to increse the temperatre, for NW or music transformer, I have achieved resonable result with tempearture = (1.4,1.1) which is a bit higher than your default
However, for mask/remix, with the same temperature, the repeating note problem shows again. Is there any specific explaination about the parameter "temperature"?

**Comment by firedie** [comment]
Moreover, I try to train on a small scale of dataset about 800 midis. They are all chords progression.s Thus I hope if I input a sequence of chords, it will generate the rest.But here is what I have got:
![image](https://user-images.githubusercontent.com/18274020/64219120-afe7ee00-cef6-11e9-8ed2-84840865a067.png)
Training take 200 epochs, with train loss 0.56 val loss 2.12 and 0.57 acc. At the geneartion procedure, I have tried different settings, still got the same result, any suggestion for that?
Do you have loss graph for your training procedure? I am not sure what is the target loss/acc to achieve a reasonable result.
Thanks!

**Comment by bearpelican** [comment]
Hey sorry for the delay. I just got around to updating the pre-trained model. You should be getting much better results for mask/remix now. 
Pretrained models [here](https://github.com/bearpelican/musicautobot#pretrained-models)

A quick explanation fo the change: the models are now specifically trained on either masked pitches or masked durations using [data augmentation](https://github.com/bearpelican/musicautobot/blob/a84a362710f879d2ea47e07bac65e7c0a6876bb1/musicautobot/multitask_transformer/dataloader.py#L99)

Accuracy:
Hmm it's a bit weird that your validation loss is so much higher than your training loss. 200 epochs is bit high, so you might be overfitting to your dataset. 
I would monitor the loss after each epoch. Or you could also use the [SaveModelCallback](https://github.com/bearpelican/musicautobot/blob/master/scripts/run_music_transformer.py#L79) so save your best validation performing model.

I usually get anywhere from 75 to 85% accuracy depending on the dataset. Are you using the MusicTransformer model or the MultitaskModel to train the chord progressions?

**Comment by firedie** [comment]
Now I am focusing on MusicTransformer. Here is my plan:
Preparing about 800 midis which only contain chord progressions(no melodys at all). If I take this midi set as training set, then the pre-train model could only focus only chord progression. However, things didn't work out. 
I have also downloaded the hook theroy dataset which contains about 35,000 midis, the pretrain model is much better than the previous one, however, it still not as good as yours (not even close).
I will try out your new model, thanks for keeping upgrading this repo!
